[Contributing] Add more information on deprecations (#12075)
symfony/symfony-docs: contributing/code/conventions.rst

.. _contributing-code-conventions-deprecations:

Deprecating code
----------------

From time to time, some classes and/or methods are deprecated in the
framework; that happens when a feature implementation cannot be changed
because of backward compatibility issues, but we still want to propose a
"better" alternative. In that case, the old implementation can simply be
**deprecated**.

Deprecations must only be introduced on the next minor version of the impacted
component (or bundle, bridge, contract).
They can exceptionally be introduced on previous supported versions if they are critical.

A new class (or interface, trait) cannot be introduced as deprecated, or
contains deprecated methods.

A feature is marked as deprecated by adding a ``@deprecated`` phpdoc to
relevant classes, methods, properties, ...::

    /**

>      * @deprecated since Symfony 2.8.

Reviewer: Why the rewording here?

cc @greg0ire

Reviewer 2: I added this after talking with @nicolasgrekas, what's the deal?

Reviewer: Yes you added the vendor/... thingy which now gets removed, that’s my point

Reviewer 2: I know, I was more asking @fancyweb or @nicolas-grekas to explain the sudden change of heart

Reviewer 3: I think the misunderstanding is because this is the "Contribute to Symfony" doc ... so it's not a generic-purpose doc and we don't need things like "vendor-name/package-name" because "vendor" is always "Symfony" and the "package-name" is not usually included in the Symfony deprecation messages.

Reviewer 4: we're now getting rid of the `to be removed` part as it doesn't bring much value
about vendors, @javiereguiluz is right, but it'd be nice to hint ppl that for other vendors, the package should really be mentionned
maybe as an rst note?

Reviewer 2: Regarding the package name, there is not always a way to guess it from a class name, for instance here:

https://github.com/symfony/var-exporter/blob/8539c2cec7202d244058075351c61aa852ffa344/Internal/Exporter.php#L120

Also, why remove the `Use XXX instead.` part? If there is a replacement, it should absolutely be mentioned IMO.

Author: The "use XXX instead" is not removed. This example is just below.

Reviewer 4: @greg0ire this link points to a notice, not to a deprecation :)

Reviewer 2: Oh right my bad, here is another example:

https://github.com/symfony/symfony/blob/38143c604f68038e28cf2fa2fbb61c04fc68b98c/src/Symfony/Component/Serializer/Encoder/CsvEncoder.php#L51

Reviewer 4: bad one indeed :)

     */

The deprecation message must indicate the version when the class was deprecated,
and whenever possible, how the feature was replaced::

    /**
     * @deprecated since Symfony 2.8, use Replacement instead.
     */

When the replacement is in another namespace than the deprecated class, its FQCN must be used::

    /**
     * @deprecated since Symfony 2.8, use A\B\Replacement instead.
     */

> A PHP ``E_USER_DEPRECATED`` error must also be triggered to help people with th migration::

Reviewer: th -> the

.. _contributing-code-conventions-deprecations:

Deprecating code
----------------

From time to time, some classes and/or methods are deprecated in the
framework; that happens when a feature implementation cannot be changed
because of backward compatibility issues, but we still want to propose a
"better" alternative. In that case, the old implementation can simply be
**deprecated**.

Deprecations must only be introduced on the next minor version of the impacted
component (or bundle, bridge, contract).
They can exceptionally be introduced on previous supported versions if they are critical.

> A new class (or interface, trait) cannot be introduced as deprecated, or

Reviewer 2: ```suggestion
A new class (or interface, or trait) cannot be introduced as deprecated, or
```

> contains deprecated methods.

Reviewer 3: `contains` -> `contain`  (because it's the continuation of the phrase, so `cannot ... contain`)

A new method cannot be introduced as deprecated.

A feature is marked as deprecated by adding a ``@deprecated`` phpdoc to
relevant classes, methods, properties, ...::

    /**
     * @deprecated since Symfony 2.8.
     */

> The deprecation message must indicate the version when the class was deprecated,

Reviewer 2: ```suggestion
The deprecation message must indicate the version in which the class was deprecated,
```

    /**
     * @deprecated since Symfony 4.4, use Symfony\Component\Routing\Loader\ContainerLoader instead.
     */
    class ServiceRouterLoader extends ObjectRouteLoader

.. _`ServiceRouterLoader`: https://github.com/symfony/symfony/blob/4.4/src/Symfony/Component/Routing/Loader/DependencyInjection/ServiceRouterLoader.php

The deprecation must be added to the ``CHANGELOG.md`` file of the impacted component::

    4.4.0
    -----

    * Deprecated the `Deprecated` class, use `Replacement` instead.

It must also be added to the ``UPGRADE.md`` file of the targeted minor version
(``UPGRADE-4.4.md`` in our example)::

    DependencyInjection
    -------------------

    * Deprecated the `Deprecated` class, use `Replacement` instead.

Finally, its consequences must be added to the ``UPGRADE.md`` file of the next major version
(``UPGRADE-5.0.md`` in our example)::

    DependencyInjection
    -------------------

    * Removed the `Deprecated` class, use `Replacement` instead.

All these tasks are mandatory and must be done in the same pull request.

Removing deprecated code
------------------------

> Removing deprecated code can only be done once every 2 year, on the next major version of the

Reviewer 2: ```suggestion
Removing deprecated code can only be done once every 2 years, on the next major version of the
```
